### DISCUSSION TITLE
Mesh Refinement for nuclei insertion in phase field mode;

### DISCUSSION THREAD

**Discussion by vermaprk** [idaholab/moose#25220] [Q&A Modules: Phase field] [answered]
I am trying to model the nucleation followed by growth by inserting nuclei at a surface. The nuclei are inserted only at a single time step (instantaneous nucleation). However, the maximum number of nuclei that I can insert is only twice the number of mesh elements at a particular time step. I tried to refine the mesh using `nucleation marker` but this is not resulting into increase in number of nuclei. Please suggest some method to increase the number of nuclei without increasing the number of mesh elements for the phase field model as this results in huge computation cost.
@dschwen @laagesen 

Thanks

**Comment by laagesen** [comment]
Hi, thanks for your question. If you are inserting all the nuclei at the same time, I don't think there is a good reason to use the nucleation system. Can't you just add them in the initial conditions for your simulation? 

Maybe you are doing this because need to simulate the evolution of your system for some time prior to the instantaneous nucleation event? Even if you need to do this, there is probably a better way than using the nucleation system. For example you could run one simulation for pre-nucleation, then start another simulation where you read in all the other necessary fields from the pre-nucleation simulation (i.e. concentrations) using a SolutionUserObject, and initialize the order parameter field with all the nuclei you want. SolutionUserObject documentation is below, with links to other necessary objects and examples.

https://mooseframework.inl.gov/source/userobjects/SolutionUserObject.html

Also be aware that no matter what you do, you are going to need to use an adequate number of elements to properly resolve each nucleus. Probably at an absolute minimum 3 elements across the interface and 5-7 across the interior of the nucleus. This is an inherent limitation of the phase-field method. So there isn't really a way to add a bunch of nuclei in to your system without also adding the mesh elements necessary to resolve them adequately.

**Reply by vermaprk** [reply]
So if my system is 100 um x 100 um and I am going to place a 0.1 um nuclei as a initial condition. Will I need around 5000 x 5000 mesh elements. That is computationally very expensive. Can't I have non uniform mesh elements being concentrated only at the nuclei ? Can I execute this using the targeted mesh refinement prior to nucleus insertion.

**Reply by GiudGiud** [reply]
no you wont. @laagesen gave you criteria near and inside the nucleus, not everywhere else.

MOOSE can handle non-uniform meshes, it can perform adaptive mesh refinement (AMR), and it can generate non-uniform meshes (this one for example: https://mooseframework.inl.gov/source/meshgenerators/CartesianMeshGenerator.html)

I think you may find in the phase field modules examples of resolving a nucleaus using AMR

**Reply by vermaprk** [reply]
Thanks for the reply. Can you suggest some mesh generator in which I can refine mesh along y-axis at a specific interval in x only. The `CartesianMeshGenerator` is refining the mesh along y-axis independently from the x-axis as per my understanding.
My system requires x intervals as 18 4 78 (xmax = 100)
with grids ix 18 40 78 (concentrated grids for the 4 interval)
But now I will require y grids for x=18 interval as iy=100 (my is ymax=100)
                                                    x= 4 interval as iy= 400 (concentrated grids along y)
                                                    x= 78 interval as iy =100
Thank you

**Reply by GiudGiud** [reply]
CMG is usually good but I can see how that doesnt work for you here. 
Ok so let's do a regular or near regular grid in the CMG
then
a SubdomainBoundingBoxGenerator around the nucleus location
then
a RefineBlockGenerator on the new block, which will locally increase the mesh density

**Reply by laagesen** [reply, marked as answer]
Before you do anything with large meshes, I would suggest running a simple 2-nucleus problem to make sure all your model parameters are settings are correct if you haven't already done so. Create 2 particles with one smaller than the other using a uniform mesh fine enough to resolve the interface, and verify that you have about 3 elements through the interface throughout the simulation, and that the small particle coarsens away and transfers its mass to the larger one. This is will allow you identify any issues prior to spending a lot of time waiting for larger simulations to run.

For a phase-field simulation, I'd recommend using the adaptivity system and letting it choose where to refine the mesh, rather than relying on specifying using a `SubdomainBoundingBoxGenerator` and `RefineBlockGenerator` - the issue is that as the microstructure changes, the initial nuclei will coarsen, so the positions where you need a finer mesh may change.

I'd recommending creating the mesh using a `GeneratedMesh` generator and using the adaptivity system to refine where needed. You can see an example of adaptivity using a two-order parameter model

/Users/aagelk/projects/moose/modules/phase_field/examples/anisotropic_interfaces/GrandPotentialPlanarGrowth.i

Please have a look at the documentation for the adaptivity system here:

https://mooseframework.inl.gov/syntax/Adaptivity/
